fix and update parts and schemas

zargot opened the pull request:
fixes #122

- add updated parts and schemas
- add gen of v1 schemas to the schema gen tool
- add xlsx validation tool
- fix a pretty bad bug where cerb rules were being repeated
- fix allowed rule
- fix gen of v1 schemas
- fix minlength rule
- fix table attribute mapping
- fix type rule
- fix v1 catsets
- fix v1 validation
- fix doc gen error due to strip of partID that isn't a string, but an int??

Trace for later investigation of doc gen error:
https://github.com/Big-Life-Lab/PHES-ODM-Validation/actions/runs/4088485603/jobs/7050208854

yulric commented on src/odm_validation/part_tables.py:
```diff
@@ -372,59 +400,106 @@ def map_ids(mappings: Dict[PartId, PartId], part_ids: List[PartId],
-        assert len(attr_table_ids) > 0
-        for id in attr_table_ids:
-            table_attrs[id].append(attr)
+    def _is_cat_v1(p: Part) -> bool:
```
A version 1 category should also have a version 1 variable

yulric reviewed (changes requested):
1. One of your commits was to remove duplicate entries in the forbidden and allowed fields in the schema. Can you explain why that was hapenning?
2. We should figure out a way to not hardcode the versions for which we need to generate validation schemas. Ideally we would get those from the firstReleased and lastUpdated columns in the dictionary
3. We need to handle multiple values (they're seperated by ;) in the the version 1 columns (version1Table, version1Variable, and version1Category)
4. I'm not sure how the tests for version 2 schemas are passing since the firstReleased and lastUpdated are set to v1 in the code by default? And we're not adding those columns in the test parts 
5. In the tests for invalid_category the following changes should be made
    1. version 1 category parts should have their status "depreciated"
    2. Their firstReleased and lastUpdated should be v1
    3. Their catSetID should be NA
6. We should refactor how the categories are being modelled for version 1 categorical attributes.

Only 3, 4, and 5 are urgent. Everything else can wait.

zargot commented:
> 1. One of your commits was to remove duplicate entries in the forbidden and allowed fields in the schema. Can you explain why that was hapenning?

The forbidden rule duplication stopped happening for some reason, so i reverted that commit.
Allowed rule duplication seems to come from wrong handling of v1 categories due to them belonging to multiple tables. I've added a FIXME for that.

> 2. We should figure out a way to not hardcode the versions for which we need to generate validation schemas. Ideally we would get those from the firstReleased and lastUpdated columns in the dictionary

I'm assuming this is something we will do in a separate PR. Can you make an issue for it?

> 3. We need to handle multiple values (they're seperated by ;) in the the version 1 columns (version1Table, version1Variable, and version1Category)

Where was this not handled? I can't remember.

> 4. I'm not sure how the tests for version 2 schemas are passing since the firstReleased and lastUpdated are set to v1 in the code by default? And we're not adding those columns in the test parts

This seems to happen due to the logic in `is_compatible` which currently only cares about that the part is active, as long as the schema version is greater than the part version. I'll see if I can fix that now.

> 5. In the tests for invalid_category the following changes should be made
>    
>    1. version 1 category parts should have their status "depreciated"
>    2. Their firstReleased and lastUpdated should be v1
>    3. Their catSetID should be NA

TODO

> 6. We should refactor how the categories are being modelled for version 1 categorical attributes.

I'm assuming you're talking about how the code structures the category data. I agree that we have to improve it.

zargot commented on src/odm_validation/schemas.py:
```diff
@@ -22,7 +24,7 @@ def init_attr_schema(attr_id: str, rule_id: str, cerb_rules: dict,
     # ['meta'] is a list with one entry per rule.
     # ['meta']['meta'] is a `Meta` with one `MetaEntry` per part.
-    inner = cerb_rules
+    inner = deepcopy(cerb_rules)
```
done

zargot commented on tools/generate_schemas.py:
```diff
@@ -15,26 +17,41 @@
-    level=logging.DEBUG,
+    level=logging.WARNING,
     handlers=[
+        logging.StreamHandler(),
```
done

zargot commented on src/odm_validation/part_tables.py:
```diff
@@ -200,16 +209,28 @@ def _parse_version1Field(part, key) -> List[str]:
 
 
-def get_mappings(part: dict, version: Version) -> Optional[List[PartId]]:
+def _normalize_key(key: Optional[str]) -> Optional[str]:
```
done

zargot commented on src/odm_validation/part_tables.py:
```diff
@@ -219,13 +240,18 @@ def get_mappings(part: dict, version: Version) -> Optional[List[PartId]]:
     return ids
 
 
+def get_part_active(part) -> bool:
```
done. I went with `is_active(part)`.

zargot commented on src/odm_validation/part_tables.py:
```diff
@@ -200,16 +209,28 @@ def _parse_version1Field(part, key) -> List[str]:
+    return key[0].lower() + key[1:]
+
+
+def _get_mappings(part: dict, version: Version) -> Optional[List[PartId]]:
```
here we need to pass `version` to determine if there is a mapping (and should be a mapping) from ODM_VERSION to `version`. If I rename it to `_get_v1_ids`, etc.,  then it doesn't make sense to pass `version` anymore, as this is supposed to determine if we are getting v1 ids or not, so I left the name the same and added a docstring instead. Is that ok?

zargot commented:
I'm waiting with any v1 category-related changes until we have cleared up how they should work. ref chat.

yulric commented on src/odm_validation/part_tables.py:
```diff
@@ -372,59 +410,93 @@ def map_ids(mappings: Dict[PartId, PartId], part_ids: List[PartId],
+        def is_cat_v1(p: Part) -> bool:
+            "Returns true if `p` is a v1-only category."
+            return (p.get(V1_TABLE) and
+                    p.get(V1_LOCATION) == 'variableCategories' and
```
variableCategories should be moved to a variable

yulric commented on src/odm_validation/part_tables.py:
```diff
@@ -372,59 +410,93 @@ def map_ids(mappings: Dict[PartId, PartId], part_ids: List[PartId],
+
+            # find a catset attr that matches
+            for attr_id in catset_data:
+                attr = all_parts[attr_id]
```
Rename to `attr_part`

yulric commented on src/odm_validation/part_tables.py:
```diff
@@ -372,59 +410,93 @@ def map_ids(mappings: Dict[PartId, PartId], part_ids: List[PartId],
+            # find a catset attr that matches
+            for attr_id in catset_data:
+                attr = all_parts[attr_id]
+                attr_tables = set(_parse_version1Field(attr, V1_TABLE))
```
Rename to `attr_v1_tables`

yulric commented on src/odm_validation/part_tables.py:
```diff
@@ -372,59 +410,93 @@ def map_ids(mappings: Dict[PartId, PartId], part_ids: List[PartId],
+            for attr_id in catset_data:
+                attr = all_parts[attr_id]
+                attr_tables = set(_parse_version1Field(attr, V1_TABLE))
+                attr_vars = set(_parse_version1Field(attr, V1_VARIABLE))
```
Rename to `attr_v1_vars`

yulric reviewed (changes requested):
1. Can we add the status field back into the test files? They should be the source of truth for tests and adding some of the columns in the code is confusing.
